### PR TITLE
Add Riak Search Message

### PR DESCRIPTION
**Pull request by hazen**
This is the Riak 1.4 legacy version of Riak Search. Yokozuna will be added post-2.0 release.

### CODE REVIEW

**Comment by rzezeski** [comment]
Riak Search (legacy) and Yokozuna use the same protobuff messages for search so this actually supports both.

**Comment by hazen** [comment]
@rzezeski Fair enough. I guess queries might work.  My point is that the YZ-specific messages are not yet supported:
- RPBYOKOZUNAINDEXGETREQ
- RPBYOKOZUNAINDEXPUTREQ
- RPBYOKOZUNAINDEXDELETEREQ
- RPBYOKOZUNASCHEMAGETREQ
- RPBYOKOZUNASCHEMAPUTREQ

File: src/messages/riak_search.c [hunk 1]
```diff
@@ -0,0 +1,424 @@
+/*********************************************************************
+ *
+ * riak_get.c: Riak C Client Legacy Search Message
```
Comment (bookshelfdave) [inline src/messages/riak_search.c]: incorrect file name in comment

File: src/messages/riak_search.c [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+        }
+        if (search_options->has_filter) {
+            srchmsg.has_filter = search_options->has_filter;
+            riak_binary_to_pb_copy(&srchmsg.filter, search_options->filter);
+        }
+        if (search_options->has_df) {
+            srchmsg.has_df = search_options->has_df;
+            riak_binary_to_pb_copy(&srchmsg.df, search_options->df);
+        }
+        if (search_options->has_df) {
```
Comment (bookshelfdave) [inline src/messages/riak_search.c]: I think you mean `search_options->has_op` here

File: src/messages/riak_search.c [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+                           char                **target,
+                           riak_int32_t         *len,
+                           riak_int32_t         *total)  {
+    riak_int32_t wrote = 0;
+
+    wrote += riak_print_bool("has_max_score", response->has_max_score, target, len, total);
+    wrote += riak_print_float("max_score", response->max_score, target, len, total);
+    wrote += riak_print_bool("has_num_found", response->has_num_found, target, len, total);
+    wrote += riak_print_int("num_found", response->num_found, target, len, total);
+    wrote += riak_print_int("num_docs", response->n_docs, target, len, total);
```
Comment (bookshelfdave) [inline src/messages/riak_search.c]: num_docs vs n_docs

File: test/cunit/test_search.c [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    field_limits.data = (riak_uint8_t*)"field_limits1";
+    field_limits.len = 13;
+    riak_search_options_add_fl(cfg, opt, &field_limits);
+    field_limits.data = (riak_uint8_t*)"field_limits2";
+    riak_search_options_add_fl(cfg, opt, &field_limits);
+    CU_ASSERT_EQUAL(riak_search_options_get_n_fl(opt), 2)
+    riak_binary **results = riak_search_options_get_fl(opt);
+    CU_ASSERT_NOT_EQUAL_FATAL(results, NULL)
+    CU_ASSERT_EQUAL(memcmp(riak_binary_data(results[0]), "field_limits1", 12), 0)
+    CU_ASSERT_EQUAL(memcmp(riak_binary_data(results[1]), "field_limits1", 12), 0)
```
Comment (bookshelfdave) [inline test/cunit/test_search.c]: should "field_limits2" be a result?

File: test/cunit/test_search.c [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    CU_ASSERT_EQUAL(riak_search_get_n_docs(response), 2);
+    CU_ASSERT_EQUAL(riak_search_get_has_max_score(response), RIAK_TRUE);
+    riak_float32_t score = riak_search_get_max_score(response);
+    CU_ASSERT(score > 0.35); // Value is 0.353553
+    CU_ASSERT(score < 0.36);
+    CU_ASSERT_EQUAL(riak_search_get_has_num_found(response), RIAK_TRUE);
+    CU_ASSERT_EQUAL(riak_search_get_num_found(response), 2);
+    riak_search_doc *docs = riak_search_get_docs(response);
+    CU_ASSERT_FATAL(docs != NULL)
+    for(int i = 0; i < riak_search_get_n_docs(response); i++) {
```
Comment (bookshelfdave) [inline test/cunit/test_search.c:306]: is this loop needed if the body is commented out?

File: test/cunit/test_search.c [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+    riak_error err = riak_config_new_default(&cfg);
+    CU_ASSERT_FATAL(err == ERIAK_OK)
+
+    err = riak_connection_new(cfg, &cxn, "localhost", "1", NULL);
+    CU_ASSERT_FATAL(err == ERIAK_CONNECT)
+    err = riak_operation_new(cxn, &rop, NULL, NULL, NULL);
+    CU_ASSERT_FATAL(err == ERIAK_OK)
+
+    riak_uint8_t bytes[] = { 0x1c,0x0a,0xda,0x01,0x0a,0x18,0x0a,0x02,0x69,
```
Comment (bookshelfdave) [inline test/cunit/test_search.c:236]: can you comment on exactly what ALLTHEBYTES are?

Comment (hazen) [inline test/cunit/test_search.c:236]: It's a pre-encoded PBC message.  I tried this approach to test the decoding.  The downside is that the message is not flexible.  I'm open to other suggestions on how to test this.

File: test/cunit/test_search.c (hunk 5 above)
Comment (hazen) [inline test/cunit/test_search.c:306]: Looks like this isn't finished.  But the test passes ;-)  Probably need to look at what the actual returned docs are.

File: test/cunit/test_search.c (hunk 6 above)
Comment (bookshelfdave) [inline test/cunit/test_search.c:236]: oh, I just meant a comment in the code mentioning that it's a pre-encoded PBC message. I thought maybe it was a gif of the Hamburglar.

Comment (hazen) [inline test/cunit/test_search.c:236]: No, actually **THIS** is a gif of the Hamburglar: hamburglar

**Comment by bookshelfdave** [comment]
:+1:

thanks for the hamburgling

**Pull request merged**